sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala

```diff
@@ -268,12 +270,26 @@ object ExtractPythonUDFs extends Rule[LogicalPlan] {
               evalTypes.mkString(","))
           }
           val evalType = evalTypes.head
+
+          val hasUDTInput = child.output.exists(
+            attr => attr.dataType.isInstanceOf[UserDefinedType[_]])
+          val hasUDTReturn = validUdfs.exists(udf => udf.dataType.isInstanceOf[UserDefinedType[_]])
+
           val evaluation = evalType match {
             case PythonEvalType.SQL_BATCHED_UDF =>
               BatchEvalPython(validUdfs, resultAttrs, child)
-            case PythonEvalType.SQL_SCALAR_PANDAS_UDF | PythonEvalType.SQL_SCALAR_PANDAS_ITER_UDF
-                 | PythonEvalType.SQL_ARROW_BATCHED_UDF =>
+            case PythonEvalType.SQL_SCALAR_PANDAS_UDF | PythonEvalType.SQL_SCALAR_PANDAS_ITER_UDF =>
               ArrowEvalPython(validUdfs, resultAttrs, child, evalType)
+            case PythonEvalType.SQL_ARROW_BATCHED_UDF =>
+
+              if (hasUDTInput || hasUDTReturn) {
+                // Use BatchEvalPython if UDT is detected
+                logWarning("Arrow optimization disabled due to UDT input or return type. " +
+                  "Falling back to non-Arrow-optimized UDF execution.")
```

Comment:
I think we should use the structured logging framework :-).

Reply:
Makes sense! Adjusted

Reply:
Adjusted thank you!